identity/app/controllers/AccountDeletionController.scala

```diff
@@ -0,0 +1,162 @@
+
+import actions.AuthenticatedActions
+import com.exacttarget.fuelsdk.ETSubscriber
+import common.ExecutionContexts
+import conf.IdentityConfiguration
+import discussion.api.DiscussionApi
+import form.Mappings
+import idapiclient.responses.CookiesResponse
+import idapiclient.{EmailPassword, IdApiClient}
+import model.{IdentityPage, NoCache}
+import play.api.data.validation._
+import play.api.data.{Form, Forms}
+import play.api.i18n.MessagesApi
+import play.api.mvc._
+import play.filters.csrf.{CSRFAddToken, CSRFCheck}
+import services._
+import utils.SafeLogging
+import scala.concurrent.Future
+import scalaz.EitherT
+import scalaz.std.scalaFuture._
+
+class AccountDeletionController(
+    api: IdApiClient,
+    idRequestParser: IdRequestParser,
+    idUrlBuilder: IdentityUrlBuilder,
+    authenticationService: AuthenticationService,
+    authenticatedActions: AuthenticatedActions,
+    val messagesApi: MessagesApi,
+    identityApiClient: IdApiClient,
+    signinService : PlaySigninService,
+    conf: IdentityConfiguration,
+    discussionApi: DiscussionApi,
+    exactTargetService: ExactTargetService)
+  extends Controller with ExecutionContexts with SafeLogging with Mappings with implicits.Forms {
+
+    import authenticatedActions._
+    import views.html.fragments.profile.deletion._
+
+    val page = IdentityPage("/delete", "Delete Account", Some("delete"))
+
+    val accountDeletionForm = Form(Forms.single("password" -> Forms.text.verifying(Constraints.nonEmpty)))
+
+    def renderAccountDeletionForm = CSRFAddToken { authActionWithUser.async { implicit request =>
+        val idRequest = idRequestParser(request)
+        val form = accountDeletionForm.bindFromFlash.getOrElse(accountDeletionForm)
+
+        selectDeletionType.map { _ match {
+            case AutoDeletion => NoCache(Ok(accountDeletion(page, idRequest, idUrlBuilder, form, Nil)))
+            case e:ManualDeletion => Ok(accountDeletionManual(page, idRequest, idUrlBuilder))
+          }
+        }
+      }
+    }
+
+    def processAccountDeletionForm = CSRFCheck { authActionWithUser.async { implicit request =>
+        val idRequest = idRequestParser(request)
+        val boundForm = accountDeletionForm.bindFromRequest
+
+        boundForm.fold(
+          formWithErrors => Future.successful(SeeOther(routes.AccountDeletionController.renderAccountDeletionForm().url).flashing(formWithErrors.toFlash)),
+          password => deleteAccount(boundForm, EmailPassword(request.user.user.primaryEmailAddress, password, None), idRequest))
+      }
+    }
+
+    def renderAccountDeletionConfirmForm = Action.async { implicit request =>
+      Future(Ok(accountDeletionConfirm(page, idRequestParser(request), idUrlBuilder)))
+    }
+
+    private def deleteAccount[A](
+        boundForm: Form[String],
+        emailPasswdAuth: EmailPassword,
+        idRequest: IdentityRequest)(implicit request: AuthenticatedActions.AuthRequest[A]) = {
+
+      sealed trait AccountDeletionFailures
```

Comment:
This (and the `processFailures` function below) looks like unnecessary indirection to me. Why not just create the response directly? This would make heaps of line noise vanish and would also mean you have the details from the API error in scope for the log message.